### PR TITLE
Added label api to mvg

### LINKED ISSUE #42: UI: Adding labeling capabilities on client side
- [ ] Mirror server labeling functions into MVG

### PR DESCRIPTION
**Pull request by OskarLiew**
Added labeling functionality to MVG

The test will not pass until labeling vikinganalytics/vibium-cloud#262 has been merged

Closes #42

### CODE REVIEW

File: mvg/mvg.py [hunk 1]
```diff
@@ -765,6 +765,111 @@ def get_analysis_results(self, request_id: str) -> dict:
 
         return response.json()
 
+    # Labels
+    def create_label(
+        self,
+        sid: str,
+        timestamp: int,
+        label: str,
```
Comment (FredrikWartenberg) [inline mvg/mvg.py:773]: This is not according to the API specification, which lacks severity and names the notes "meta" which also is a dictionary.
I think the severity should be part of "meta" as for labels not related to failures severity does not make any sense.

File: mvg/mvg.py [hunk 2]
```diff
@@ -765,6 +765,111 @@ def get_analysis_results(self, request_id: str) -> dict:
...
+            Optional notes for the label, by default ""
+        """
+        logger.info("endpoint %s", self.endpoint)
+        logger.info(f"Creating label for {sid} - {timestamp}")
+
+        label_data = {"label": label, "severity": severity, "notes": notes}
+
+        self._request("post", f"/sources/{sid}/labels/{timestamp}", json=label_data)
+
+    def get_label(self, sid: str, timestamp: int) -> dict:
```
Comment (FredrikWartenberg) [inline mvg/mvg.py:799]: This is not according to the API spec
- takes a list of timestamps
- (as what I can see), the timestamp argument is required, which is not according to specification

File: mvg/mvg.py [hunk 3]
```diff
@@ -765,6 +765,111 @@ def get_analysis_results(self, request_id: str) -> dict:
...
+            label information
+        """
+
+        logger.info("endpoint %s", self.endpoint)
+        logger.info("Getting label")
+
+        response = self._request("get", f"/sources/{sid}/labels/{timestamp}")
+        return response.json()
+
+    def list_labels(self, source_id: str) -> List[dict]:
```
Comment (FredrikWartenberg) [inline mvg/mvg.py:821]: This should be part of the non-argument version of get label.

File: mvg/mvg.py [hunk 4]
```diff
@@ -765,6 +765,111 @@ def get_analysis_results(self, request_id: str) -> dict:
...
+        logger.info("endpoint %s", self.endpoint)
+        logger.info("Getting labels")
+
+        response = self._request("get", f"/sources/{source_id}/labels")
+        return response.json()
+
+    def update_label(
```
Comment (FredrikWartenberg) [inline mvg/mvg.py:828]: No according to specification (see above)

File: mvg/mvg.py [hunk 5]
```diff
@@ -42,26 +43,25 @@ def __init__(self, endpoint: str, token: str):
 
         Raises
         ------
-        HTTPError
+        ConnectionError
```
Comment (FredrikWartenberg) [inline mvg/mvg.py:46]: Should be a separate PR.

File: tests/test_api_sources.py [hunk 6]
```diff
@@ -59,6 +59,23 @@ def session(vibium):
     return session
```
Comment (FredrikWartenberg) [inline tests/test_api_sources.py:61]: Should be a separate PR

File: tests/test_api_sources.py [hunk 7]
```diff
@@ -265,44 +282,118 @@ def test_sources_cru_existing(session):
...
+    session.delete_measurement(tabular_source, ts0)
     with pytest.raises(HTTPError) as exc:
-        session.read_single_measurement(SOURCE_ID_TABULAR, ts0)
+        session.read_single_measurement(tabular_source, ts0)
     assert exc.value.response.status_code == 404
 
 
+def test_create_label(session, tabular_source_with_measurements):
+    timestamps = tabular_dict["timestamp"]
+    label1 = {"label": "normal", "severity": 0, "notes": ""}
```
Comment (FredrikWartenberg) [inline tests/test_api_sources.py:326]: naming should not use label both for dict and actual label,

File: tests/test_api_sources.py [hunk 8]
```diff
@@ -265,44 +282,118 @@ def test_sources_cru_existing(session):
...
+    label2 = {"label": "failure", "severity": 100, "notes": "This is really bad!"}
+
+    session.create_label(tabular_source_with_measurements, timestamps[0], **label1)
```
Comment (FredrikWartenberg) [inline tests/test_api_sources.py:329]: Using keyword arguments here is a sign that the method signature is too complex, rather stick to the API spec (see my comment on "severity" above.

File: tests/test_api_sources.py [hunk 9]
```diff
@@ -265,44 +282,118 @@ def test_sources_cru_existing(session):
...
+    session.update_label(
+        tabular_source_with_measurements, timestamps[0], "normal", 0, "It wasn't so bad"
+    )
+
+    label1 = session.get_label(tabular_source_with_measurements, timestamps[0])
+
+    assert label1 == {"label": "normal", "severity": 0, "notes": "It wasn't so bad"}
+
+
+def test_update_label(session, tabular_source_with_measurements):
```
Comment (FredrikWartenberg) [inline tests/test_api_sources.py:364]: duplicate function?

File: tests/test_mvg.py [hunk 10]
```diff
@@ -3,12 +3,12 @@
...
-from requests import HTTPError
 import pytest
+import requests
 from mvg import MVG
 
 
 def test_create_session_without_server():
     """We should get an HTTPError because we cannot connect to localhost:8000"""
-    with pytest.raises(HTTPError):
+    with pytest.raises(requests.ConnectionError):
```
Comment (FredrikWartenberg) [inline tests/test_mvg.py:13]: see above (other PR)

**Review by FredrikWartenberg — CHANGES_REQUESTED** [review]
### General
- The code does not conform to the API specification (` c0000f0 2021-06-18 | Add workflows section and update labeling section`)
- "severity" and "notes" argument should be moved into the "meta" dict.
- especially severity should not be a mandatory argument, as there can be labels where a severity value does not make any sense.
- I general I think the design approach in the API spec with meta for transporting opaque data between client and server side is important to be adhered to, as we do not know what kind of information our customers want to attach to labels (or sources or measurements)

### Minor (no need to act)
- some changes should have been separate PRs.

File: mvg/mvg.py (hunk 1 above)
Comment (tuix) [reply mvg/mvg.py:773]: When there is no issue with the label, severity level is zero. Severity level is always mandatory because we want to be able to create health index by taking expected value of the severity. It is mandatory there not part of meta data.

**Review by FredrikWartenberg — APPROVED** [review]
OK with changes after internal discussion about actual version of API spec.

**Comment by FredrikWartenberg** [comment]
Merged after running additional basic integration test `pytest tests --host http://127.0.0.1:8000` locally
